cpp/include/cudf/column/column_device_view_base.cuh

```diff
@@ -0,0 +1,782 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/detail/offsets_iterator.cuh>
+#include <cudf/fixed_point/fixed_point.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/types.hpp>
+#include <cudf/utilities/bit.hpp>
+#include <cudf/utilities/traits.hpp>
+
+#include <cuda/std/optional>
+#include <cuda/std/type_traits>
+
+#include <algorithm>
+#include <type_traits>
+
+/**
+ * @file column_device_view_base.cuh
+ * @brief Column device view class definitions
+ */
+
+namespace CUDF_EXPORT cudf {
+
+/**
+ * @brief Indicates the presence of nulls at compile-time or runtime.
+ *
+ * If used at compile-time, this indicator can tell the optimizer
+ * to include or exclude any null-checking clauses.
+ *
+ * @ingroup column_classes
+ *
+ */
+struct nullate {
+  struct YES : cuda::std::true_type {};
+  struct NO : cuda::std::false_type {};
+  /**
+   * @brief `nullate::DYNAMIC` defers the determination of nullability to run time rather than
+   * compile time. The calling code is responsible for specifying whether or not nulls are
+   * present using the constructor parameter at run time.
+   */
+  struct DYNAMIC {
+    DYNAMIC() = delete;
+    /**
+     * @brief Create a runtime nullate object.
+     *
+     * @see cudf::column_device_view::optional_begin for example usage
+     *
+     * @param b True if nulls are expected in the operation in which this
+     *          object is applied.
+     */
+    constexpr explicit DYNAMIC(bool b) noexcept : value{b} {}
+    /**
+     * @brief Returns true if nulls are expected in the operation in which this object is applied.
+     *
+     * @return `true` if nulls are expected in the operation in which this object is applied,
+     * otherwise false
+     */
+    CUDF_HOST_DEVICE constexpr operator bool() const noexcept { return value; }
+    bool value;  ///< True if nulls are expected
+  };
+};
+
+namespace detail {
+/**
+ * @brief An immutable, non-owning view of device data as a column of elements
+ * that is trivially copyable and usable in CUDA device code.
+ *
+ * column_device_view_base and derived classes do not support has_nulls() or
+ * null_count().  The primary reason for this is that creation of column_device_views
+ * from column_views that have UNKNOWN null counts would require an on-the-spot, and
+ * not-obvious computation of null count, which could lead to undesirable performance issues.
+ * This information is also generally not needed in device code, and on the host-side
+ * is easily accessible from the associated column_view.
+ */
+class alignas(16) column_device_view_base {
+ public:
+  // TODO: merge this offsets column index with `strings_column_view::offsets_column_index`
+  static constexpr size_type string_offsets_column_index{0};  ///< Child index of the offsets column
```

Comment:
We could consider changing this to just `offsets_column_index` since the offsets index for lists column is the same --
would help make this look less specific to strings perhaps. 
There is very little chance this value will change in the future but it is always good to have a name for a constant like this..

Reply:
done. thanks!